README.md
<p align="center">
    <img src="https://i.imgur.com/xnQyZaU.png" height="200">
</p>


# Sony Bravia

This is a plugin for [Homebridge](https://github.com/nfarina/homebridge) to control your **Sony Bravia Android TV**. 

This plugin supports following functions:

- **Power Switch** (on/off)
- **Inputs** like HDMI, Scart, CEC Devices, AV, WIFI, DVB:T, DVB:C etc.
- **Apps** like YouTube, Prime Video etc.
- **Channels:** Your favourite channels as inputs.
- **Remote control:** native iOS Remote control
- **Login** with PSK or without PSK (for older models)
- **WOL** support Wake on Lan

## Installation instructions

After [Homebridge](https://github.com/nfarina/homebridge) has been installed:

-  ```sudo npm i -g git://github.com/Hypovenom0180/homebridge-sony-bravia.git```

## Preparing the TV with PSK (Prefered method)

- Set **Remote start** to **ON** _(Settings -> Network -> Remote Start)_
- Change **Authentication** to **Normal and Pre-Shared Key** _(Settings -> Network -> IP Control -> Authentication)_
- Enter a **Pre-Shared Key** _(Settings -> Network -> IP control -> Pre-Shared Key)_


## Preparing the TV without PSK

- Set **Remote start** to **ON** _(Settings -> Network -> Remote Start)_
- Change **Authentication** to **Normal** _(Settings -> Network -> IP Control -> Authentication)_


## Basic Configuration

 ```
 "platforms": [
    {
      "tvs": [
        {
          "name": "Lounge TV",
          "ip": "192.168.1.254",
          "psk": "loungetvpsk"
        }
      ],
      "platform": "SonyBravia"
    }
  ]
 ```

You can also configure the plugin through the Homebridge UI. At this stage, only some options are avaliable to configure via the Homebridge UI.
 
 
 ## Options

| **Attributes** | **Required** | **Usage** |
|------------|----------|-------|
| name | **Yes** | **Unique Name** for the TV Accessory.   |
| ip | **Yes** | IP adress from your Sony Bravia Android TV |
| mac | **No** | MAC address from TV (required if using WOL!) |
| port | **No** | If you have problems with connecting to the TV, try a different port _(Default: 80)_ |
| psk | **No** | Your PRE SHARED KEY _(see preparing the TV above)_ |
| extraInputs | **No** | Inputs for "Scart, Composite, Wifidisplay" _(Default: false)_ |
| cecInputs | **No** | Inputs for connected cec devices like Apple TV _(Default: false)_ |
| channelInputs | **No** | An Array of Channel input types (DVBT/DVBC/DVBS/ANALOG) _(Default: false)_ |
| channels | **No** | List of your favourite channels (channel numbers from tv and source) to display these as inputs in the TV accessory _(Default: false)_ |
| apps | **No** | List of your favourite apps to display as inputs in the TV accessory  _(Default: false)_ |
| wol | **No** | Wake On Lan  _(Default: false)_ |
| customSpeaker | **No** | If true, an extra speaker accessory will be published to HomeKit  _(Default: false)_ |
| speakerType | **No** | Type of the custom speaker (switch, lightbulb, speaker) _(Default: speaker)_ |
| interval | **No** | Polling interval in seconds _(Default: 10s)_ |


## External Accessories

Please note that all TVs will be published as "external accessories"! That means, you need to add these TVs manually to HomeKit. It will be not exposed via the bridge. This is necessary due to handling of the TVs via Apple Homekit.


## Contributing

You can contribute to this homebridge plugin in following ways:

- Report issues.
- Contribute bug fixes.
- Contribute changes to extend the capabilities.

## Credit

A Huge thank you to SeydX and their homebridge-bravia-tvos plugin! This plugin is essentially their plugin with a few changes made by bkonings, niou-ns, and me that enables support for Homebridge 1.0 and Config UI X.
https://github.com/SeydX/homebridge-bravia-tvos#readme


Changelog.md
# Changelog

## 3.1.0
- All TVs are now external accessories
- Updated README
- General tidy up

## 3.0.5
- Bugfixes
- Use input label when defined [(#47)](https://github.com/SeydX/homebridge-bravia-tvos/pull/47)


## 3.0.3
- Bugfixes
- Debugging improvements


## 3.0.2
- Debugging improvements


## 3.0.1
- Bugfixes


## 3.0
- Completely rewritten code
- Faster
- Bugfixes and Improvements
- Better communication
- More memory efficient
- Designed for iOS 12.2
- Debug mode

## 2.0
- Dynamic Platform Plugin
- Bugfixes
- Performance improvements
- Ability to add multiple TVs
- Designed for iOS 12.1 and lower

## 1.0 ([homebridge-bravia-tv](https://github.com/SeydX/homebridge-bravia-tv))
- Init release


lib/Bravia.js
'use strict';

const axios = require('axios');
const wol = require('wake_on_lan');
const tcpp = require('tcp-ping');
const debug = require('debug')('BraviaPlatformApi');
const base64 = require('base-64');

const tcpping = opt => new Promise((resolve, reject) => tcpp.ping(opt, (err, data) => err ? reject(err) : resolve(data)));
const tcpprobe = (ip,port) => new Promise((resolve, reject) => tcpp.probe(ip,port, (err, available) => err ? reject(err) : resolve(available)));

const timeout = ms => new Promise(res => setTimeout(res, ms));

const isObject = function(a) {
  return (!!a) && (a.constructor === Object);
};

class Bravia {
  constructor(platform, context, waitUntilAuthenticated) {
  
    if(!context.ip) throw 'No ip found in config!';
    
    this.ip = context.ip;
    this.port = context.port || 80;
    this.name = context.name;
    this.context = context;
    
    this.logger = platform.logger;
   
    this.opt = {
      address: this.ip,
      port: this.port,
      timeout: 3000,
      attempts: 500
    };
    
    this.url = 'http://' + this.ip + ':' + this.port;
    
    this.header = {
      'Content-Type': 'application/json'
    };
    
    if(context.psk){
    
      this.logger.info(this.name + ': Initalizing Bravia API with PSK');
    
      this.header['X-Auth-PSK'] = context.psk;
      this.authenticated = true;
      
    } else {
      
      this.HBpath = platform.api.user.storagePath();   
      this.store = require('json-fs-store')(this.HBpath);
      
      if(!waitUntilAuthenticated){
        this.logger.info(this.name + ': Initalizing Bravia API with IP Control');
        this.checkAuthFile();
      } else {
        this.checkUntilAuthenticated();
      }
      
    }
    
    this.count = 0;
    
  }
  
  async checkUntilAuthenticated(){
  
    const self = this;
  
    let name = this.name.includes('Speaker') ? this.name.split(' Speaker')[0] : this.name;
    
    this.store.load(name, async function(err, object){
      if(err) {
      
        //self.logger.error(err);
        
        await timeout(5000);
        self.checkUntilAuthenticated();
      
      } else {

        //check registration
          
        if(object.cookie && object.expires){
  
          const expireDate = new Date(object.expires).getTime();
          let today = new Date();
          today.setDate(today.getDate() + 1);
          today = today.getTime();
    
          if( (expireDate - today) <= 0 ){
    
            await timeout(5000);
            self.checkUntilAuthenticated();
    
          } else {
    
            //everything ok!
            self.header.Cookie = object.cookie;
            self.authenticated = true;
    
          }
  
        } else {
        
          await timeout(5000);
          self.checkUntilAuthenticated();
        
        }
      
      }

    });
  
  }
  
  getAuth(){
    return new Promise((resolve, reject) => {
      this._handleAuth((err, res) => {
        if(err) reject(err);
      
        resolve(res);
      
      });
    });
  }
  
  _handleAuth(callback){
    if(this.authenticated){
      callback(null, 'Authenticated');
    } else {
      this.authCount++;
  
      if(this.authCount >= 12){
  
        callback(this.name + ': Can not authenticate!', null);
        
      } else {
 
        setTimeout(this._handleAuth.bind(this,callback),5000); 
  
      }
    }
  }
  
  async checkAuthFile(){
  
    const self = this;
  
    this.store.load(self.name, async function(err, object){
      if(err) {
      
        let auth = {
          id: self.name,
          pin: 1234
        };
           
        self.store.add(auth, err => {
          if(err) throw err;
          self.checkAuthFile();
        });
      
      } else {

        //check registration
          
        if(object.cookie && object.expires){
  
          const expireDate = new Date(object.expires).getTime();
          let today = new Date();
          today.setDate(today.getDate() + 1);
          today = today.getTime();
    
          if( (expireDate - today) <= 0 ){
    
            //refresh cookie und expire
            self.checkAuthentication(null, null, true);
    
          } else {
    
            //everything ok!
            self.header.Cookie = object.cookie;
            self.authenticated = true;
    
          }
  
        } else {
  
          if(object.pin && (object.pin !== 1234)){
    
            //create cookie and expires
            self.checkAuthentication(object.pin);
    
          } else {
          
            self.checkAuthentication();
    
          }
  
        }
      
      }

    });
  
  }
  
  async checkAuthentication(user, cookie, refresh){
  
    const self = this;
    
    if(!user && !cookie && !refresh)
      this.logger.info(self.name + ': Trying to register on ' + this.ip);
  
    try {
    
      let res = await this.registerCall(user, cookie, refresh);
    
      if(res.data.error && res.data.error.includes(40005)){
          
        //Display off!
        //Turn on display for init auth!
        
        if(!this.wolTried && this.context.mac){
          this.logger.info(self.name + ': Turning on TV for init authentication...');
            
          this.wolTried = true;  
            
          await this.setPowerStatusWOL(this.context.mac);
        } else {
          if(!this.context.mac){
            this.logger.warn(self.name + ': Can not turn on TV via WOL (no mac address in config.json). Please manually turn on the TV!');
          } else {
            this.logger.warn(self.name + ': Can not turn on TV via WOL. Please manually turn on the TV!');
          }
        }
            
        await timeout(15000);
          
        this.checkAuthentication(user, cookie, refresh);
          
      } else {
    
        if(res.headers['set-cookie']){
      
          //cookie received in header!
          //something like this:
    
          /*
           * headers: { 
           *   'content-type': 'application/json',
           *   'content-length': '20',
           *   'connection': 'close',
           *   'set-cookie': [ 
           *     'auth=C78436B5E8CB9A3C6A7DD2B29DE3386A012535A3; 
           *      Path=/sony/; 
           *      Max-Age=1209600; // <- in seconds, need to refresh before it expires!
           *      Expires=Fr., 26 Apr. 2019 21:42:48 GMT+00:00' 
           *   ] 
           * }
          */
          
          let cookieHeader = res.headers['set-cookie'][0].split(';');
          
          cookieHeader.forEach( str => {
  
            if(str.includes('auth')){
              self.header.Cookie = str;
            }
    
            if(str.includes('Expires')){
              self.expires = str.split('=')[1];
            }
          
          });
          
          self.authenticated = true;
    
          this.store.load(self.name, async function(err, object){
            
            let auth = object;
            
            auth.cookie = self.header.Cookie;
            auth.expires = self.expires;
       
            self.store.add(auth, function(err){
            
              if(err) throw err; // err if JSON parsing failed
              
              if(!object.cookie)
                self.logger.info(self.name + ': Authentication successfull!');

              self.logger.info(self.name + ': ' + self.name + '.json refreshed with new cookie!');
              
            });
             
          });
      
        } else {
      
          //no cookie in header!
        
          let error = {
            status: res.status,
            message: res.statusMessage,
            data: res.data,
            headers: res.headers
          };
          
          this.logger.error(JSON.stringify(error));
      
        }
          
      }
      
    } catch(err) {
    
      if(err.response && err.response.status === 401){
      
        //not authorized yet!
        //need to register device
      
        //PIN should be appeared on screen (401)
        //start authentication
      
        this.logger.info(self.name + ': You should see a dialog with a 4-digit PIN.');
        this.logger.info(self.name + ': Please enter the four-digit PIN in ' + self.name + '.json! Located at ' + self.HBpath);
        this.logger.info(self.name + ': Trying again in 30 seconds...');
               
        await timeout(30000);
            
        this.checkAuthFile();
      
      } else {
    
        //something went wrong!
          
        /*let error = {
          code: error.response.status,
          message: error.response.statusMessage,
          data: error.response.data
        };*/
        
        this.logger.error(err);
    
      }
    
    }
  
  }
  
  async registerCall(user, cookie, refresh) {
  
    const self = this;
  
    return new Promise((resolve, reject) => {
    
      let header = {};
    
      if(!refresh){
      
        if(user){
          let basic = 'Basic ' + base64.encode(':' + user);
          header.Authorization = basic;
        }
        
        if(cookie)
          header.Cookie = cookie;
        
      }
      
      var data = '{"method":"actRegister","params":[{"clientid":"BraviaTVOS:481b0a56-fcdf-4b13-bef7-dc22acf965ab","nickname":"BraviaTVOS (homebridge-sony-bravia)","level":"private"},[{"value":"yes","function":"WOL"}]],"id":8,"version":"1.0"}';
      
      axios({
        url: self.url + '/sony/accessControl',
        method: 'POST',
        data: data,
        headers: header
      }).then(response => {
        resolve(response);
      }).catch(err => {
        reject(err);
      });
    
    });
    
  }
  
  _refreshToken() {
  
    const self = this;

    return new Promise((resolve, reject) => {
    
      if(self.context.psk){
  
        self.header['X-Auth-PSK'] = self.context.psk;
        resolve(true);

      } else {
  
        let name = self.name.includes('Speaker') ? self.name.split(' Speaker')[0] : self.name;

        this.store.load(name, async function(err, object){
    
          if(err) reject(err);

          //check registration
          
          if(object.cookie && object.expires){
  
            const expireDate = new Date(object.expires).getTime();
            let today = new Date();
            today.setDate(today.getDate() + 1);
            today = today.getTime();
    
            if( (expireDate - today) <= 0 ){
    
              //refresh cookie und expire
              self.checkAuthentication(null, null, true);
              resolve(true);
    
            } else {
    
              resolve(true);
    
            }
      
          }

        });

      } 
    });
  }

  async apiCall(url, method, version, params, req) {
  
    const self = this;
  
    let pingError;
  
    if(!await tcpprobe(this.ip,this.port)){
    
      this.logger.warn(this.name + ': TV currently offline, trying to connect (this can take up to 5 minutes) ...');
    
      try {
      
        await tcpping(this.opt);
        await timeout(10000);
        
        this.logger.info(this.name + ': Connection established!');
      
      } catch(err){
      
        pingError = true;
      
      }
      
    }    
  
    req = req ? req : 'post';   
    params = params ? [params] : [];
    
    this.count++;
    let currCount = this.count;
    debug('[Bravia Debug]: api request ' + currCount + ': ' + req + ' ' + url + ' - ' + method + (params.length ? ' - ' + JSON.stringify(params) : ''));
 
    return new Promise((resolve, reject) => {
      if(self.url && (self.header['X-Auth-PSK'] || self.header.Cookie) && !pingError){
      
        this._refreshToken().then(() => {
    
          let data;
          let header = this.header;
        
          if(url === '/sony/IRCC'){
    
            data = '<?xml version="1.0"?><s:Envelope xmlns:s="http://schemas.xmlsoap.org/soap/envelope/" s:encodingStyle="http://schemas.xmlsoap.org/soap/encoding/"><s:Body><u:X_SendIRCC xmlns:u="urn:schemas-sony-com:service:IRCC:1"><IRCCCode>' + method + '</IRCCCode></u:X_SendIRCC></s:Body></s:Envelope>';

            header['SOAPACTION'] = '"urn:schemas-sony-com:service:IRCC:1#X_SendIRCC"';
            header['Content-Type'] = 'text/xml';
            header['Content-Length'] = Buffer.byteLength(data);
    
          } else {
    
            header = this.header;
    
            data = {
              'id':1,
              'method':method,
              'version':version,
              'params':params
            };
    
            data = JSON.stringify(data);
    
          }
      
          axios({
            url: self.url + url,
            method: req,
            data: data,
            headers: header
          }).then(response => {
    
            let result;
    
            if(response.data.result){
    
              if(response.data.result.length){
                result = response.data.result[0];
              } else {
                result = 'Ok';
              }
            
              debug('[Bravia Debug]: api request ' + currCount + ': Ok - ' + JSON.stringify(result)); 
    
            } else if(response.data.results){
          
              if(response.data.results.length){
                result = response.data.results[0];
              } else {
                result = 'Ok';
              }
            
              debug('[Bravia Debug]: api request ' + currCount + ': Ok - ' + JSON.stringify(result)); 
          
            } else if(response.data.error) {
          
              if(response.data.error.includes('Illegal State')){
        
                result = 'App State';
        
                debug('[Bravia Debug]: api request ' + currCount + ': Ok - ' + result);
      
              } else if(response.data.error.includes('Display Is Turned off')){
        
                result = 'Display Off';
        
                debug('[Bravia Debug]: api request ' + currCount + ': Ok - ' + result);
        
              } else {
        
                debug('[Bravia Debug]: api request ' + currCount + ': Error - ' + JSON.stringify(response.data.error)); 
    
                reject(response.data.error); 
        
              }
    
            } else {
          
              if((isObject(response.data)||typeof response.data === 'string'||response.data instanceof String) && response.data.includes('<?xml version="1.0"?>')){
        
                result = 'Ok';
        
                debug('[Bravia Debug]: api request ' + currCount + ': Ok - ' + JSON.stringify(response.data).replace(/\s/g, ''));
        
              } else {
            
                let error = {
                  status: response.status,
                  message: response.statusText,
                  config: response.config,
                  data: response.data
                };
        
                debug('[Bravia Debug]: api request ' + currCount + ': Error Unknown');
    
                reject(error);
        
              }
    
            }
    
            resolve(result);
          }).catch(err => {
          
            let error = err;
        
            if(err.response){
              error = {
                status: err.response.status,
                message: err.response.statusText,
                config: err.config,
                data: err.response.data
              };
            } 
            
            debug('[Bravia Debug]: api request ' + currCount + ': Error Unknown');  
        
            reject(error);
          });
        
        
        }).catch(err => {
          debug('[Bravia Debug]: api request ' + currCount + ': Error!');
          reject(err);  
        });
        
      } else {
        if(!self.url){
          reject('No URL');
        } else if(!self.header['X-Auth-PSK'] || !self.header.Cookie){
          reject('No PSK/Cookie!');
        } else {
          reject('Can not connect to TV!');
        }
      }
    });
  }
  
  wolCall(mac){
  
    return new Promise((resolve, reject) => {
    
      debug('[Bravia Debug]: WOL: send magic packets to ' + mac);
    
      wol.wake(mac, function(error) {
        if (error) {
          debug('[Bravia Debug]: WOL: Error: ' + JSON.stringify(error));
          reject(JSON.stringify(error));
        } else {
          debug('[Bravia Debug]: WOL: Ok');
          resolve('Done');
        }
      });
  
    });  
  
  }
  
  //# /sony/guide
  
  getServiceProtocols(){
    return this.apiCall('/sony/guide', 'getServiceProtocols', '1.0');
  }
  
  getSupportedApiInfo(){
    return this.apiCall('/sony/guide', 'getSupportedApiInfo', '1.0');
  }

  //# /sony/system
  
  getColorKeysLayout() {
    return this.apiCall('/sony/system', 'getColorKeysLayout', '1.0');
  }
  
  getLEDIndicatorStatus() {
    return this.apiCall('/sony/system', 'getLEDIndicatorStatus', '1.0');
  }
  
  getCurrentTime() {
    return this.apiCall('/sony/system', 'getCurrentTime', '1.0');
  }
  
  getDateTimeFormat() {
    return this.apiCall('/sony/system', 'getDateTimeFormat', '1.0');
  }
  
  getNetworkSettings(type) {
    return this.apiCall('/sony/system', 'getNetworkSettings', '1.0', {'netif': type});
  }
  
  getPowerSavingMode() {
    return this.apiCall('/sony/system', 'getPowerSavingMode', '1.0');
  }
  
  getPowerStatus() {
    return this.apiCall('/sony/system', 'getPowerStatus', '1.0');
  }
  
  getRemoteControllerInfo() {
    return this.apiCall('/sony/system', 'getRemoteControllerInfo', '1.0');
  }
  
  getSystemInformation() {
    return this.apiCall('/sony/system', 'getSystemInformation', '1.0');
  }
  
  getInterfaceInformation() {
    return this.apiCall('/sony/system', 'getInterfaceInformation', '1.0');
  }
  
  getSystemSupportedFunction() {
    return this.apiCall('/sony/system', 'getSystemSupportedFunction', '1.0');
  }
  
  getWolMode() {
    return this.apiCall('/sony/system', 'getWolMode', '1.0');
  }
  
  setWolMode(state) {
    if(state===undefined) return 'No state defined!';
    return this.apiCall('/sony/system', 'setWolMode', '1.0', {enabled: state});
  }
  
  setPowerStatus(state) {
    if(state===undefined) return 'No state defined!';
    return this.apiCall('/sony/system', 'setPowerStatus', '1.0', {status: state});
  }
  
  setPowerStatusWOL(mac) {
    if(mac===undefined) return 'No mac address defined!';
    return this.wolCall(mac);
  }
  
  //# /sony/videoScreen
  
  getAudioSourceScreen() {
    return this.apiCall('/sony/videoScreen', 'getAudioSourceScreen', '1.0');
  }
  
  getPipSubScreenPosition() {
    return this.apiCall('/sony/videoScreen', 'getPipSubScreenPosition', '1.0');
  }
  
  //# /sony/audio
  
  getVolumeInformation() {
    return this.apiCall('/sony/audio', 'getVolumeInformation', '1.0', '1.1');
  }
  
  setAudioMute(state) {
    if(state===undefined) return 'No state defined!';
    return this.apiCall('/sony/audio', 'setAudioMute', '1.0', {status: state});
  }
  
  // target = headphone|speaker - volume: 1-100
  setAudioVolume(target, volume) {
    target = target === undefined ? 'speaker' : target;
    volume = volume === undefined ? 10 : volume;
    return this.apiCall('/sony/audio', 'setAudioVolume', '1.0', {target: target, volume: volume.toString()});
  }
  
  //# /sony/avContent
  
  // example: source = extInput:hdmi
  getContentCount(source) {
    if(source===undefined) return 'No source defined!';
    return this.apiCall('/sony/avContent', 'getContentCount', '1.1', {source: source});
  }
  
  // source = usb:recStorage|tv:dvbc,tv:dvbt
  getContentList(source, cnt, stIdx) {
    if(source===undefined) return 'No source defined!';
    return this.apiCall('/sony/avContent', 'getContentList', '1.2', {source: source, cnt: cnt||200, stIdx: stIdx||0});
  }
  
  getCurrentExternalInputsStatus() {
    return this.apiCall('/sony/avContent', 'getCurrentExternalInputsStatus', '1.0');
  }
  
  getParentalRatingSettings() {
    return this.apiCall('/sony/avContent', 'getParentalRatingSettings', '1.0');
  }
  
  getPlayingContentInfo() {
    return this.apiCall('/sony/avContent', 'getPlayingContentInfo', '1.0');
  }
  
  getSchemeList() {
    return this.apiCall('/sony/avContent', 'getSchemeList', '1.0');
  }
  
  // scheme = exInput|fav|tv|usb
  getSourceList(scheme) {
    if(scheme===undefined) return 'No scheme defined';
    return this.apiCall('/sony/avContent', 'getSourceList', '1.0', {scheme: scheme});
  }
  
  // e.g uri = extInput:hdmi?port=1
  setPlayContent(uri) {
    if(uri===undefined) return 'No uri defined';
    return this.apiCall('/sony/avContent', 'setPlayContent', '1.0', {uri: uri});
  }
  
  //# /sony/recording
  
  getRecordingStatus() {
    return this.apiCall('/sony/recording', 'getRecordingStatus', '1.0');
  }
  
  getSupportedRepeatType() {
    return this.apiCall('/sony/recording', 'getSupportedRepeatType', '1.0');
  }
  
  //# /sony/appControl
  
  getApplicationList() {
    return this.apiCall('/sony/appControl', 'getApplicationList', '1.0');
  }
  
  getApplicationStatusList() {
    return this.apiCall('/sony/appControl', 'getApplicationStatusList', '1.0');
  }
  
  setActiveApp(uri) {
    if(uri===undefined) return 'No uri defined';
    return this.apiCall('/sony/appControl', 'setActiveApp', '1.0', {uri: uri});
  }
  
  terminateApps() {
    return this.apiCall('/sony/appControl', 'terminateApps', '1.0');
  }
  
  //# endpoints available with 'getServiceProtocols'
  
  getMethodTypes(endpoint, versions){
    if(endpoint===undefined) return 'Endpoint required!';
    if(versions===undefined) versions = '1.0';
    return this.apiCall(`/sony/${endpoint}`, 'getMethodTypes', '1.0', versions);
  }
  
  getVersions(endpoint){
    if(endpoint===undefined) return 'Endpoint required!';
    return this.apiCall(`/sony/${endpoint}`, 'getVersions', '1.0');
  }
  
  //# IRCC
  
  setIRCC(command) {
    if(command===undefined) return 'No command defined!';
    return this.apiCall('/sony/IRCC', command);
  }
  
}

module.exports = Bravia;


src/platform.js
'use strict';

const packageFile = require('../package.json');
const LogUtil = require('../lib/LogUtil.js');
const Bravia = require('../lib/Bravia.js');
const debug = require('debug')('BraviaPlatform');

//Accessory
const Device = require('./accessory.js');
const Speaker = require('./speaker_accessory.js');

const pluginName = 'homebridge-sony-bravia';
const platformName = 'SonyBravia';

var Accessory, Service, Characteristic, UUIDGen;

module.exports = function (homebridge) {

  Accessory = homebridge.platformAccessory;
  Service = homebridge.hap.Service;
  Characteristic = homebridge.hap.Characteristic;
  UUIDGen = homebridge.hap.uuid;
  
  return SonyBravia;
};

function SonyBravia (log, config, api) {
  if (!api || !config) return;

  // HB
  this.log = log;
  this.logger = new LogUtil(null, log);
  this.debug = debug;
  this.accessories = [];
  this._accessories = new Map();
  this._devices = new Map();
  this.config = config;
  
  this.config.interval = this.config.interval * 1000||10000;                    
  this.config.tvs = Array.isArray(this.config.tvs) ? this.config.tvs : [];

  if (api) {
  
    if (api.version < 2.2) {
      throw new Error('Unexpected API version. Please update your homebridge!');
    }
    
    this.api = api;
      
    this.api.on('didFinishLaunching', this._initPlatform.bind(this));
  }
}

SonyBravia.prototype = {

  _initPlatform: function(){
  
    if(!this.config.tvs.length){
    
      this._addOrRemoveDevice();
      
    } else {
    
      this._confTVs();

    }
  
  },
  
  _confTVs: function(){
  
    this.config.tvs.map( tv => {
      this._devices.set(tv.name, tv);
    
      if(tv.customSpeaker)
        this._devices.set(tv.name + ' Speaker', tv);
    
    });
    
    this.config.tvs.map( tv => {
    
      this._addOrRemoveDevice(tv);
    
      if(tv.customSpeaker){
        tv.name = tv.name + ' Speaker';
        this._addOrRemoveDevice(tv);
      }
      
    });
    
  },

  _addOrRemoveDevice: function(object) {
  
    this.accessories.map( accessory => {

      if(!this._devices.has(accessory.displayName)){

        this._accessories.delete(accessory.displayName);
        this.removeAccessory(accessory);

      }
      
    });

    
    if(object){
    
      const accessory = this._accessories.get(object.name);

      if(!accessory){

        this._accessories.set(object.name, object);
        this.addAccessory(object);

      }
    
    }

  },
  
  addAccessory: function(object){
    
    let external = true;
    if(!external||object.name.includes('Speaker')){
      this.logger.info('Adding new accessory: ' + object.name);
    } else {
      if(!object.name.includes('Speaker'))
        this.debug('[Bravia Debug]: Adding new accessory: ' + object.name);
    }
    
    let catagory;
    
    if(object.name.includes('Speaker')){
      switch(object.speakerType){
      
        case 'lightbulb':
          catagory = 5;
          break;
          
        case 'switch':
          catagory = 8;
          break;
          
        case 'speaker':
          catagory = 26;
          break;
          
        default:
          object.speakerType = 'speaker';
          catagory = 26;
      
      }
    } else {
      catagory = 31;
    }
    
    let uuid = UUIDGen.generate(object.name);
    let accessory = new Accessory(object.name, uuid, catagory);

    accessory.context = {};
    
    this.accessories.push(accessory);

    this._addOrConfigure(accessory, object, true, external);

  },
  
  _addOrConfigure: function(accessory, object, add, external){

    this._refreshContext(accessory, object, add);    
    this._AccessoryInformation(accessory);

    this.config.tvs.map( tv => {
      if((tv.name === accessory.displayName||(tv.name + ' Speaker' === accessory.displayName && accessory.context.customSpeaker)) && tv.ip){
        if(!add) this.logger.info('Configuring accessory ' + accessory.displayName);
        
        if(accessory.displayName.includes('Speaker')){
          new Speaker(this, accessory, add);
        } else {
          new Device(this, accessory, add, external);
        }
      } else {
        if(!tv.ip) this.logger.warn('No IP defined in config.json for ' + accessory.displayName + '. Skipping...');
      }
    });

  },
  
  _refreshContext: function(accessory, object, add){
  
    accessory.reachable = true;
    accessory.context.interval = this.config.interval;
  
    if(add){ 

      accessory.context.name = object.name;
      accessory.context.ip = object.ip;
      accessory.context.mac = object.mac;
      accessory.context.port = object.port || 80;
      accessory.context.psk = object.psk;
      accessory.context.extraInputs = object.extraInputs || false;
      accessory.context.cecInputs = object.cecInputs || false;
      accessory.context.channelInputs = object.channelInputs || [];
      accessory.context.channels = object.channels || [];
      accessory.context.apps = object.apps || [];
      accessory.context.commands = object.commands || [];
      accessory.context.wol = object.wol || false;
      accessory.context.customSpeaker = object.customSpeaker || false;
      accessory.context.speakerType = object.speakerType || false;
      accessory.context.Bravia = object.Bravia;
      
      if(!object.name.includes('Speaker'))
        accessory.context.Bravia = new Bravia(this, object);
    
    } else {
    
      this.config.tvs.map( tv => {
    
        if(accessory.displayName === tv.name || accessory.displayName === tv.name + ' Speaker'){
    
          accessory.context.name = accessory.displayName.includes('Speaker') ? tv.name + ' Speaker' : tv.name;
          accessory.context.ip = tv.ip;
          accessory.context.mac = tv.mac;
          accessory.context.port = tv.port || 80;
          accessory.context.psk = tv.psk;
          accessory.context.extraInputs = tv.extraInputs || false;
          accessory.context.cecInputs = tv.cecInputs || false;
          accessory.context.channelInputs = tv.channelInputs || [];
          accessory.context.channels = tv.channels || [];
          accessory.context.apps = tv.apps || [];
          accessory.context.commands = tv.commands || [];
          accessory.context.wol = tv.wol || false;
          accessory.context.customSpeaker = tv.customSpeaker || false;
          accessory.context.speakerType = tv.speakerType || false;
          
          if(accessory.displayName === tv.name)
            accessory.context.Bravia = new Bravia(this, tv);
    
        }
    
      });

    }
    
  },
  
  _AccessoryInformation: function(accessory){

    let serial = accessory.displayName.includes('Speaker') ? 'S-' + accessory.context.ip.replace(/\./g, '') : accessory.context.ip.replace(/\./g, '');
  
    accessory.getService(Service.AccessoryInformation)
      .setCharacteristic(Characteristic.Name, accessory.displayName)
      .setCharacteristic(Characteristic.Identify, accessory.displayName)
      .setCharacteristic(Characteristic.Manufacturer, 'Hypovenom0180')
      .setCharacteristic(Characteristic.Model, 'Sony')
      .setCharacteristic(Characteristic.SerialNumber, serial)
      .setCharacteristic(Characteristic.FirmwareRevision, packageFile.version);
  
  },

  configureAccessory: function(accessory){

    this._accessories.set(accessory.displayName, accessory);  
    
    this.accessories.push(accessory);
    this._addOrConfigure(accessory);
  
  },

  removeAccessory: function (accessory) {
    if (accessory) {

      this.logger.warn('Removing accessory: ' + accessory.displayName + '. No longer configured.');

      let newAccessories = this.accessories.map( acc => {
        if(acc.displayName !== accessory.displayName){
          return acc;
        }
      });

      let filteredAccessories = newAccessories.filter(function (el) {
        return el != null;
      });

      this.api.unregisterPlatformAccessories(pluginName, platformName, [accessory]); 

      this.accessories = filteredAccessories;

    }
  }

};
